src/communication/network_interface_sr.c

```diff
@@ -10223,3 +10226,190 @@ ssession_stop_attached_threads (void *session)
 {
   session_stop_attached_threads (session);
 }
+
+void
+scdc_get_lsa (THREAD_ENTRY * thread_p, unsigned int rid, char *request, int reqlen)
+{
+  OR_ALIGNED_BUF (OR_INT_SIZE + OR_LOG_LSA_ALIGNED_SIZE) a_reply;
+  char *reply = OR_ALIGNED_BUF_START (a_reply);
+  char *ptr;
+  LOG_LSA start_lsa;
+  time_t input_time;
+  uint64_t b_start_lsa;
+  int error;
+
+  ptr = or_unpack_int64 (request, &input_time);
+
+  error = cdc_get_lsa (thread_p, input_time, &start_lsa);
+
+  ptr = or_pack_int (reply, error);
+  ptr = or_pack_log_lsa (ptr, &start_lsa);
+  (void) css_send_data_to_client (thread_p->conn_entry, rid, reply, OR_ALIGNED_BUF_SIZE (a_reply));
+
+  return;
+}
+
+void
+scdc_set_configuration (THREAD_ENTRY * thread_p, unsigned int rid, char *request, int reqlen)
+{
+  OR_ALIGNED_BUF (OR_INT_SIZE) a_reply;
+  char *reply = OR_ALIGNED_BUF_START (a_reply);
+  char *ptr;
+  int error;
+  int max_log_item, time_out, all_in_cond, num_user, num_class;
+  uint64_t *classoids = NULL;
+  char **user = NULL;
+  int user_len = 0;
+
+  char *dummy_user = NULL;
+  int timeout;
+
+  ptr = or_unpack_int (request, &max_log_item);
+  ptr = or_unpack_int (ptr, &timeout);
+  ptr = or_unpack_int (ptr, &all_in_cond);
+  ptr = or_unpack_int (ptr, &num_user);
+
+  if (num_user > 0)
+    {
+      user = (char **) malloc (sizeof (char *) * num_user);
+      if (user == NULL)
+	{
+	  error = ER_OUT_OF_VIRTUAL_MEMORY;
+	  goto end;
+	}
+      for (int i = 0; i < num_user; i++)
+	{
+	  ptr = or_unpack_string (ptr, &dummy_user);
```

Comment:
How about 'ptr = or_unpack_string (ptr, &user[i]);'?

If you do that, I think you don't need to call the memcpy() and db_private_free_and_init().

Reply:
db_private_alloc() is called in or_unpack_string() and returns allocated memory. 
the memory allocated with db_private_alloc() is required to be freed before end of the function. But I need to keep memory allocated until CDC is finished. That's why I can't use or_unpack_string directly.